aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp

```diff
@@ -1537,6 +1537,142 @@ void fake_quant_grad_per_channel_cpu(TensorIterator &iter, int64_t quant_min, in
     });
 }
 
+template <typename T>
+void quantized_layer_norm_kernel_impl(
+    const Tensor& X,
+    const Tensor& gamma,
+    const Tensor& beta,
+    int64_t M,
+    int64_t N,
+    float eps,
+    Tensor* Y) {
+
+  using qVec = vec256::Vec256<T>;
+  using fVec = vec256::Vec256<float>;
+
+  TORCH_INTERNAL_ASSERT(X.numel() == M * N, "Unexpected num elements in X");
+  TORCH_INTERNAL_ASSERT(!gamma.defined() || gamma.numel() == N,
+      "Unexpected size of gamma");
+  TORCH_INTERNAL_ASSERT(!beta.defined() || beta.numel() == N,
+      "Unexpected size of beta");
+  T* X_data = X.data_ptr<T>();
+  const float* gamma_data = gamma.defined() ? gamma.data_ptr<float>() : nullptr;
+  const float* beta_data = beta.defined() ? beta.data_ptr<float>() : nullptr;
+  T* Y_data = Y->data_ptr<T>();
+  const float c = 1.0f / static_cast<float>(N);
+  const bool gamma_null = gamma_data == nullptr;
+  const bool beta_null = beta_data == nullptr;
+
+  int64_t x_zp = X.q_zero_point();
+  float x_scale = X.q_scale();
+
+  fVec x_zp_vec = fVec((float)x_zp);
+  fVec one_vec = fVec(1.0f);
+  fVec zero_vec = fVec(0.0f);
+
+  float x_fake_scale = 1.0f;
+  fVec x_fake_scale_vec = fVec(x_fake_scale);
+  fVec x_fake_scale_zp_neg_premul_vec = x_fake_scale_vec * x_zp_vec.neg();
+
+  int64_t y_zp = Y->q_zero_point();
+  float y_scale = Y->q_scale();
+  float y_inv_scale = 1.0f / y_scale;
+
+  // 8 floats in a 256 bit Vec256
+  constexpr int kFloatVLen = 8;
+  // N ints in a qVec
+  int64_t kIntVLen = kFloatVLen * qVec::float_num_vecs();
+  // portion of layer that can be vectorized
+  int64_t kNumIntVecInLayer = N / kIntVLen;
+  // remainder of layer that cannot be vectorized
+  int64_t kNonVecRemInLayer = N % kIntVLen;
+
+  at::parallel_for(0, M, 1, [&](int64_t start, int64_t end) {
+    for (int64_t i = start; i < end; ++i) {
+
+      T* X_ptr = X_data + i * N;
+      T* Y_ptr = Y_data + i * N;
+
+      // First pass: calculate mean and variance.
+
+      int64_t layerSumShifted = 0.0f;
+      int64_t layerSumSquaresShifted = 0.0f;
+      for (int64_t vecIdx = 0; vecIdx < kNumIntVecInLayer; vecIdx++) {
+        int64_t vecStartIdx = vecIdx * kIntVLen;
+        auto qXVec = qVec::loadu(X_ptr + vecStartIdx);
+        layerSumShifted += qXVec.h_sum();
+        layerSumSquaresShifted += qXVec.h_sum_sq();
```

Comment:
Instead of horizontal reduction on each vector (We are executing many instruction for reducing each vector), something like https://github.com/pytorch/FBGEMM/blob/master/src/OptimizedKernelsAvx2.cc#L19-L31 might be faster.